### PR TITLE
`CodeBlock`: spike—CodeMirror version (HDS-2277)

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary

This PR is intended to be an experimental code spike for implementing a `CodeBlock` component using **[CodeMirror](https://www.npmjs.com/package/codemirror)**. It's not necessarily intended to be merged...

- **Showcase:** https://hds-showcase-git-hds-2277-code-block-spike-hashicorp.vercel.app/components/code-block

👉 **Compare to the PrismJs version:** https://github.com/hashicorp/design-system/pull/1603

### :hammer_and_wrench: Detailed description

Implemented features:

- Syntax highlighting for different coding languages
  - for custom themes we could modify CSS from a pre-set theme to fit our needs
- Copy code button option
- Line numbers option
- Line wrapping option
- Max-height option (no toggle implemented)
- Solution for displaying different languages using `Hds::Tabs` with multiple `CodeBlock` components
- Editable / readOnly option

TODO:
 - Highlight lines (Odd/even striping would be easy to implement but highlighting specific lines would be more involved)
 - Toggle for maxHeight (Would depend on desired design though)

### :link: External links

- Jira ticket: [HDS-2277](https://hashicorp.atlassian.net/browse/HDS-2277)

***

### 👀 Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed
- [ ] Confirm that A11y tests have been run locally for this component

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-2277]: https://hashicorp.atlassian.net/browse/HDS-2277?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/addon/components/hds/code-block/index.hbs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{{!
+  Copyright (c) HashiCorp, Inc.
+  SPDX-License-Identifier: MPL-2.0
+}}
+
+<div class="hds-code-block-wrapper">
```
Comment (didoo) [inline packages/components/addon/components/hds/code-block/index.hbs:6]: should we use `<pre><code>` here, as I've seen done in some implementations?

File: packages/components/addon/components/hds/code-block/index.hbs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  <div
+    class={{this.classNames}}
+    {{style maxHeight=@maxHeight}}
+    {{hds-code-mirror
+      value=@value
+      options=@options
+      lineNumbers=@lineNumbers
+      lineWrapping=@lineWrapping
+      readOnly=@readOnly
```
Comment (didoo) [inline packages/components/addon/components/hds/code-block/index.hbs:15]: probably the `readonly` use case is the most common one (so should be `true` by default)?

File: packages/components/addon/modifiers/hds-code-mirror.js
Comment (didoo) [inline packages/components/addon/modifiers/hds-code-mirror.js:1]: what is the relation of this file, vs the typescript one? this is a leftover?

**Review by didoo — COMMENTED** [review]
Left a few preliminary comments

File: packages/components/addon/components/hds/code-block/index.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  // }
+
+  /**
+   * Get the class names to apply to the component.
+   * @method classNames
+   * @return {string} The "class" attribute to apply to the component.
+   */
+  get classNames() {
+    let classes = ['hds-code-block'];
+
```
Comment (didoo) [inline packages/components/addon/components/hds/code-block/index.js:22]:
```suggestion
    let classes = ['hds-code-block'];

    classes.push('hds-typography-code-200');

```

File: packages/components/addon/components/hds/code-block/index.hbs (hunk 2 above)
Comment (KristinLBradley) [reply packages/components/addon/components/hds/code-block/index.hbs:15]: Probably so, I can change it.

File: packages/components/addon/modifiers/hds-code-mirror.js
Comment (KristinLBradley) [reply packages/components/addon/modifiers/hds-code-mirror.js:1]: Oops, I included a duplicate typescript version by accident. I've deleted it now.

File: packages/components/addon/components/hds/code-block/index.hbs (hunk 1 above)
Comment (KristinLBradley) [reply packages/components/addon/components/hds/code-block/index.hbs:6]: CodeMirror generates a bunch of `div`s with `<pre>` tags nested inside for each line of code.

The `code` tag is only supposed to be used to wrap phrasing content, not block-level content such as a `div`.

We could look at this more closely when we implement the actual component however.

File: packages/components/addon/modifiers/hds-code-mirror.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import { registerDestructor } from '@ember/destroyable';
+import codemirror from 'codemirror';
+
+// import type Owner from '@ember/owner';
+
+// import 'core/utils/register-codemirror-hcl';
+import 'codemirror/mode/javascript/javascript';
+import 'codemirror/mode/go/go';
+import 'codemirror/mode/markdown/markdown';
+import 'codemirror/mode/shell/shell';
```
Comment (KristinLBradley) [inline packages/components/addon/modifiers/hds-code-mirror.js:13]: I'm been trying to think about how to enable consumers to choose what language modes they import but I really don't know how that could work. Do either of you have any recommendations for how that could possibly work or resources I could use to research? @didoo @alex-ju

Comment (didoo) [reply packages/components/addon/modifiers/hds-code-mirror.js:13]: probably I would start by looking at how it's done in the consumers' codebases: how do they do it?

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    </SG.Item>
+
+    <SG.Item {{style width="calc(50% - 1rem)"}} @label="with lineWrapping=true">
+      <Hds::CodeBlock
+        @lineWrapping={{true}}
+        @value="console.log('I am JavaScript code', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam');"
+      />
+    </SG.Item>
+  </Shw::Flex>
+
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:44]:
```suggestion
```

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  {{!-- <Shw::Text::h3>with lineWrapping=true</Shw::Text::h3>
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:45]:
```suggestion
```

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:46]:
```suggestion
```

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  <Hds::CodeBlock
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:47]:
```suggestion
```

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    @lineWrapping={{true}}
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:48]:
```suggestion
```

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    @value="console.log('I am JavaScript code', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam');"
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:49]:
```suggestion
```

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+  /> --}}
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:50]:
```suggestion
```

File: packages/components/addon/modifiers/hds-code-mirror.js (hunk 4 above)
Comment (KristinLBradley) [reply packages/components/addon/modifiers/hds-code-mirror.js:13]: Haven't found examples of consumers doing this. They seem to just specify the individual languages they need as part of their components.

File: packages/components/addon/modifiers/hds-code-mirror.js
Comment (KristinLBradley) [inline packages/components/addon/modifiers/hds-code-mirror.js:1]: The starting code for this follows Cloud-ui's implementation: https://github.com/hashicorp/cloud-ui/blob/main/addons/core/addon/modifiers/code-mirror.ts

File: packages/components/addon/modifiers/hds-code-mirror.js (hunk 4 above)
Comment (didoo) [reply packages/components/addon/modifiers/hds-code-mirror.js:13]: @KristinLBradley maybe there are interesting code snippets here?
https://github.com/search?q=%22codemirror%2Fmode%22+user%3Ahashicorp&type=code&ref=advsearch

Comment (KristinLBradley) [reply packages/components/addon/modifiers/hds-code-mirror.js:13]: Thanks!

**Pull request closed without merging**